posts/2018-08-25-http4s-error-handling-mtl.md

```diff
@@ -0,0 +1,327 @@
+---
+layout: post
+title: Http4s error handling with Cats Meow MTL
+
+meta:
+  nav: blog
+  author: gvolpe
+  pygments: true
+
+tut:
+  scala: 2.12.6
+  binaryScala: "2.12"
+  dependencies:
+    - org.scala-lang:scala-library:2.12.6
+    - org.typelevel::cats-core:1.1.0
+    - org.typelevel::cats-effect:1.0.0-RC2
+    - org.http4s::http4s-blaze-server:0.19.0-M1
+    - org.http4s::http4s-circe:0.19.0-M1
+    - org.http4s::http4s-dsl:0.19.0-M1
+    - io.circe::circe-core:0.10.0-M1
+    - io.circe::circe-generic:0.10.0-M1
+    - co.fs2::fs2-core:1.0.0-M1
+
+---
+
+As a longtime `http4s` user I keep on learning new things and I'm always trying to come up with the best practices for writing http applications. This time I want to talk about my latest achievements in error handling within the context of an http application where it basically means mapping each business error to the appropiate [http response](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes).
+
+So let's get started by putting up an example of an http application with three different endpoints that interacts with a `UserAlgebra` that may or may not fail with some specific errors.
+
+If you are one of those who don't like to read and prefer to jump straight into the code please find it [here](https://gist.github.com/gvolpe/3fa32dd1b6abce2a5466efbf0eca9e94) :)
+
+### User Algebra
+
+We have a simple `UserAlgebra` that let us perform some actions such as finding and persisting users.
+
+```tut:book:silent
+case class User(username: String, age: Int)
+case class UserUpdateAge(age: Int)
+
+trait UserAlgebra[F[_]] {
+  def find(username: String): F[Option[User]]
+  def save(user: User): F[Unit]
+  def updateAge(username: String, age: Int): F[Unit]
+}
+```
+
+And also an ADT of the possible errors that may arise. I'll explain later in this post why it extends `Throwable`.
+
+```tut:book:silent
+sealed trait UserError extends Throwable
+case class UserAlreadyExists(username: String) extends UserError
+case class UserNotFound(username: String) extends UserError
+case class InvalidUserAge(age: Int) extends UserError
+```
+
+### User Interpreter
+
+And here we have a simple interpreter for our `UserAlgebra` for demonstration purposes so you can have an idea on how the logic would look like. In a real-life project an interpreter will more likely have some `Repository` as a dependency instead of using a `Ref`.
+
+```tut:book:silent
+import cats.effect.Sync
+import cats.effect.concurrent.Ref
+import cats.syntax.all._
+
+object UserInterpreter {
+
+  def create[F[_]](implicit F: Sync[F]): F[UserAlgebra[F]] =
+    Ref.of[F, Map[String, User]](Map.empty).map { state =>
+      new UserAlgebra[F] {
+        private def validateAge(age: Int): F[Unit] =
```

Comment:
I normally think of validators as returning the validated value on the happy path and probably would have returned an `F[Int]`.  But since you're not transforming it or lifting it into a newtype, I guess it doesn't really matter.

Reply:
I think I get your point but since it's just a simple interpreter for the user to have an idea of how the business logic would look like I didn't want to get into `validation`. I think leaving it as it is wouldn't be that bad since I really want the users to focus on the error handling topic instead.